File: 8-string-to-integer-atoi/string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        int n = s.length();
        int t = 0;
        long ans = 0;
        bool key = true;
        int sign = 1;
        // Just to remove the leading white spaces:
        while (t < n && s[t] == ' ') {
            t++;
        }

        //Checking thr signs:
        if (s[t] == '-' || s[t]=='+') {
            sign = (s[t]=='-') ? -1 : 1;
            t++; 
        }

        // Skipping the leading zeroes:
        while (t < n and s[t] == '0') {
            t++;
        }

        // Checking whether the character is a non-integer or not?
        while (t < n and isdigit(s[t])) {
            ans = ans * 10 + (s[t] - '0');
            if (ans * sign >= INT_MAX)
                return INT_MAX;
            if (ans * sign <= INT_MIN)
                return INT_MIN;

            t++;
        }

        return ans * sign;
    }
};

File: 8-string-to-integer-atoi/string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        # 1. Check for the whitespaces:
        # 2. Check for the sign
        # 3. check for the leading zeroes
        # 4. Check for the non-digit characters
        # 5. stop untill the end or overfow happens
        n = len(s)

        def solve(index, sign, ans):
            if index == 0:
                while index < n and s[index] == " ":
                    index += 1

                if index < n and (s[index] == "+" or s[index] == "-"):
                    sign = -1 if s[index] == "-" else 1
                    index += 1

                while index < n and s[index] == "0":
                    index += 1

            if index >= n or not s[index].isdigit():
                return ans * sign

            ans = ans * 10 + int(s[index])
            index += 1
            if sign == 1 and ans > 2**31 - 1:
                return 2**31 - 1
            if sign == -1 and ans > 2**31:
                return -(2**31)
            return solve(index, sign, ans)

        return solve(0, 1, 0)
